### PR TITLE
Simplify `ProtocolContext` and `ConsensusContext` creation

### PR DESCRIPTION
**Pull request by fab-10**
## PR description

Built on top of https://github.com/hyperledger/besu/pull/7791, #7862 and #7863

There are no functional changes in this PR, since it is refactor only. The goal is to simplify the ProtocolContext creation, 
Moreover the ConsensusContext is now created before the ProtocolContext, removing the need of the factory.

### Thanks for sending a pull request! Have you done the following?

- [ ] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [ ] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [ ] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [ ] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [ ] unit tests: `./gradlew build`
- [ ] acceptance tests: `./gradlew acceptanceTest`
- [ ] integration tests: `./gradlew integrationTest`
- [ ] reference tests: `./gradlew ethereum:referenceTests:referenceTests`

### CODE REVIEW

**Comment by daniellehrner** [comment]
@fab-10 Could we use a builder pattern instead? `createPartiallyInitialized` improves the readability, but I still don't know which parts are initialized and which are still missing.

The builder could hide the multi step creation from the user.

**Comment by fab-10** [comment]
> @fab-10 Could we use a builder pattern instead? `createPartiallyInitialized` improves the readability, but I still don't know which parts are initialized and which are still missing.
> 
> The builder could hide the multi step creation from the user.

I thought about it, but in this case it is not possible since `ProtocolContext` is being used, before it is completely set, to create what it is missing, the synchronizer, so in order to use a builder we should make the objects that are needed to create the synchronizer use the `ProtocolContext.Builder` but it does not seem a good idea.
Maybe at the end the synchronizer should not be in the `ProtocolContext` at all, to break this circular dependency, will check what implies removing the synchronizer from the `ProtocolContext`

File: besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java [hunk 1]
```diff
@@ -724,6 +725,7 @@ public BesuController build() {
     }
 
     protocolContext.setSynchronizer(synchronizer);
+    protocolContext.seal();
```
[the remark this answers is not in the document]
Comment (fab-10) [reply besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java:728]: thanks @garyschulte for reviewing, but this part has been superseded by this new PR https://github.com/hyperledger/besu/pull/7862  to remove the synchronizer from `ProtocolContext`

**Review by daniellehrner — APPROVED** [review]
LGTM

**Pull request merged**